### ISSUE TITLE
DatePicker Typing in date has odd behaviour

### ISSUE THREAD

**Issue by JoshuaC3** [jupyter-widgets/ipywidgets#1322]
If I try to type 01012017 into the DatePicker I get the following:

![image](https://cloud.githubusercontent.com/assets/11645712/25530114/f6bff7c0-2c1c-11e7-8315-9ab8a07e8fed.png)

It all goes well up until the 0 in 2017. The does the reset on the day and month to dd mm and the 20 at the start of the year to 00.

I can go back, once 2017 is typed and retype 0101

**Comment by JoshuaC3** [comment]
I will add that this problem occurred before the ipywidget 7.0 update

**Comment by jasongrout** [comment]
I'm seeing the same issue on basically the current alpha of 7.0. The input does seem to work on a plain html page. Setting as sprint-friendly because it *should* be straightforward to track down...

**Comment by pbugnion** [comment]
As far as I can tell, the problem goes away if one removes conversions of the date to/from UTC. 

Instead of this:

```javascript
    private _update_value() {
        var value = this.model.get('value');
        this._datepicker.valueAsDate = createDateAsUTC(value);
    }

    private _picker_change() {
        this.model.set('value', convertDateToUTC(this._datepicker.valueAsDate));
        this.touch();
    }
```

having this:

```javascript
    private _update_value() {
        var value = this.model.get('value');
        this._datepicker.valueAsDate = value;
    }

    private _picker_change() {
        this.model.set('value', this._datepicker.valueAsDate);
        this.touch();
    }
```

I think that the underlying problem with `convertDateToUTC` is caused by the Javascript Date constructor having slightly different meanings attached to two digit years. For instance, in the chrome console:

```javascript
> var d = new Date("0005-04-28")
> d
Thu Apr 28    5 01:00:00 GMT+0100 (BST)
> d.getFullYear()
5
> new Date(d.getUTCFullYear(), d.getUTCMonth(), d.getUTCDay())
Tue Apr 04 1905 00:00:00 GMT+0100 (BST) // Check the year!!
```

I presume the rationale for converting the dates to and from UTC is to avoid having to pass timezone information to the backend? 

I can see the following possible fixes:

 - Also serialize timezone information in the JSON object that gets serialised and pass this into the datetime constructor in Python.
 - instead of passing a date that's decomposed into elements, pass either an ISO string representation or a number of millis since the epoch. This is nice since (as far as I can tell), both of these minimise the amount of date time manipulation we have to do ourselves.

I'm happy to implement either options, but it'd be useful to get some feedback on which you guys think is best?

**Comment by jasongrout** [comment]
> I think that the underlying problem with convertDateToUTC is caused by the Javascript Date constructor having slightly different meanings attached to two digit years.

Ah, very interesting! Good catch!

> I presume the rationale for converting the dates to and from UTC is to avoid having to pass timezone information to the backend?

I think that's what the motivation was, i.e., our dates were timezone unaware, so we make all dates with reference to UTC.

The problem seems to happen when a partial date is specified when typing. A partial date is used because the change handler acts on the change event (https://github.com/jupyter-widgets/ipywidgets/blob/657c582820ec45dabaa0c34c671d7a267eb93a99/jupyter-js-widgets/src/widget_date.ts#L118). Since it doesn't make sense to sync a partially-typed date, why don't we only update the value when the input is blurred (change the `change` event to a `blur` event), or maybe also when enter is pressed (a `keypress` event handler checking for keycode 13)?

That's orthogonal to changing how we sync dates. @SylvainCorlay - do you have a preference for changing how we sync dates too?

**Comment by pbugnion** [comment]
Fixing this properly (i.e. validating the entire behaviour of the widget) is a bit tricky until issue #1320 is addressed.

**Comment by jasongrout** [comment]
This is fixed in #1348. Thanks @JoshuaC3 for bringing it up, and @pbugnion for fixing it!
